presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java

```diff
@@ -0,0 +1,180 @@
+import com.facebook.presto.sql.analyzer.TypeSignatureProvider;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.spi.StandardErrorCode.FUNCTION_NOT_FOUND;
+import static com.facebook.presto.spi.StandardErrorCode.NOT_FOUND;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class FunctionManager
+{
+    private final FunctionNamespace globalFunctionNamespace;
+    private volatile Map<SqlPathElement, FunctionNamespace> dynamicFunctionNamespaces;
+    private final TypeManager typeManager;
+    private final BlockEncodingSerde blockEncodingSerde;
+    private final FeaturesConfig featuresConfig;
+
+    public FunctionManager(TypeManager typeManager, BlockEncodingSerde blockEncodingSerde, FeaturesConfig featuresConfig)
+    {
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
+        this.featuresConfig = requireNonNull(featuresConfig, "featuresConfig is null");
+        this.globalFunctionNamespace = new StaticFunctionNamespace(typeManager, blockEncodingSerde, featuresConfig);
+        this.dynamicFunctionNamespaces = ImmutableMap.of();
+    }
+
+    public synchronized void addFunctions(String catalog, String schema, List<? extends SqlFunction> functions)
+    {
+        if (!dynamicFunctionNamespaces.containsKey(catalog)) {
+            FunctionNamespace namespace = new StaticFunctionNamespace(typeManager, blockEncodingSerde, featuresConfig);
+            dynamicFunctionNamespaces = ImmutableMap.<SqlPathElement, FunctionNamespace>builder()
+                    .putAll(dynamicFunctionNamespaces)
+                    .put(new SqlPathElement(catalog, schema), namespace)
+                    .build();
+        }
+        dynamicFunctionNamespaces.get(catalog).addFunctions(functions);
+    }
+
+    public List<SqlFunction> listFunctions(Session session)
+    {
+        ImmutableList.Builder<SqlFunction> builder = ImmutableList.builder();
+        builder.addAll(globalFunctionNamespace.listFunctions());
+        for (SqlPathElement element : session.getPath().getParsedPath()) {
+            FunctionNamespace namespace = dynamicFunctionNamespaces.get(resolveSqlPathElement(element, session));
+            if (namespace != null) {
+                builder.addAll(namespace.listFunctions());
+            }
+        }
+        return builder.build();
+    }
+
+    public Signature resolveFunction(Session session, QualifiedName name, List<TypeSignatureProvider> parameterTypes)
+    {
+        try {
+            return globalFunctionNamespace.resolveFunction(name, parameterTypes);
+        }
+        catch (Exception e) {
```

Comment:
As you pointed out offline, I also don't like the way we do resolution through catching exceptions. We should get rid of this.